pf/internal/plugin/provider_server.go

```diff
@@ -16,702 +16,80 @@ package plugin
-			case pl.DiffUpdate:
-				kind = pulumirpc.PropertyDiff_UPDATE
-			case pl.DiffUpdateReplace:
-				kind, replaces = pulumirpc.PropertyDiff_UPDATE_REPLACE, append(replaces, path)
-			}
-
-			detailedDiff[path] = &pulumirpc.PropertyDiff{
-				Kind:      kind,
-				InputDiff: diff.InputDiff,
-			}
-		}
-	}
-
-	return &pulumirpc.DiffResponse{
-		Replaces:            replaces,
-		DeleteBeforeReplace: diff.DeleteBeforeReplace,
-		Changes:             changes,
-		Diffs:               diffs,
-		DetailedDiff:        detailedDiff,
-	}, nil
-}
-
-type forwardServer struct {
-	plugin.UnimplementedProvider
-
-	parameterize func(context.Context, plugin.ParameterizeRequest) (plugin.ParameterizeResponse, error)
-}
-
-func (p forwardServer) Parameterize(
-	ctx context.Context, req plugin.ParameterizeRequest,
-) (plugin.ParameterizeResponse, error) {
-	return p.parameterize(ctx, req)
-}
-
-func (p *providerServer) Parameterize(
-	ctx context.Context, req *pulumirpc.ParameterizeRequest,
-) (*pulumirpc.ParameterizeResponse, error) {
-	return plugin.NewProviderServer(&forwardServer{
-		parameterize: p.provider.ParameterizeWithContext,
-	}).Parameterize(ctx, req)
-}
-
-func (p *providerServer) GetSchema(ctx context.Context,
-	req *pulumirpc.GetSchemaRequest,
-) (*pulumirpc.GetSchemaResponse, error) {
-	var subpackageVersion *semver.Version
-	if v := req.GetSubpackageVersion(); v != "" {
-		ver, err := semver.Parse(v)
-		if err != nil {
-			return nil, fmt.Errorf("invalid SubpackageVersion: %w", err)
-		}
-		subpackageVersion = &ver
-	}
-	schema, err := p.provider.GetSchemaWithContext(ctx, plugin.GetSchemaRequest{
-		Version:           int(req.GetVersion()),
-		SubpackageName:    req.SubpackageName,
-		SubpackageVersion: subpackageVersion,
-	})
-	if err != nil {
-		return nil, err
-	}
-	return &pulumirpc.GetSchemaResponse{Schema: string(schema)}, nil
-}
-
-func (p *providerServer) GetPluginInfo(ctx context.Context, req *pbempty.Empty) (*pulumirpc.PluginInfo, error) {
-	info, err := p.provider.GetPluginInfoWithContext(ctx)
-	if err != nil {
-		return nil, err
-	}
-	return &pulumirpc.PluginInfo{Version: info.Version.String()}, nil
-}
-
-func (p *providerServer) Attach(ctx context.Context, req *pulumirpc.PluginAttach) (*pbempty.Empty, error) {
-	// NewProviderServer should take a GrpcProvider instead of Provider, but that's a breaking change
-	// so for now we type test here
-	if grpcProvider, ok := p.provider.(interface {
-		Attach(address string) error
-	}); ok {
-		err := grpcProvider.Attach(req.GetAddress())
-		if err != nil {
-			return nil, err
-		}
-		return &pbempty.Empty{}, nil
-	}
-	// Else report this is unsupported
-	return nil, status.Error(codes.Unimplemented, "Attach is not yet implemented")
+	p := &providerServer{
+		ResourceProviderServer: plugin.NewProviderServer(
+			providerThunk{
+				configencoding.New(configEncoding, NewProvider(provider)),
+			},
+		),
+	}
+
+	// This is a hack to work around limitations in the Pulumi protocol handshake:
```

Comment:
Is this hack new? We need to add tests for this if so.

Reply:
The hack is new but the behavior is already under test. Removing the hack fails tests.